superset-frontend/webpack.proxy-config.js

```diff
@@ -0,0 +1,171 @@
+ */
+const fs = require('fs');
+const zlib = require('zlib');
+const path = require('path');
+// eslint-disable-next-line
+const parsedArgs = require('minimist')(process.argv.slice(2));
+
+const { supersetPort = 8088, superset: supersetUrl = null } = parsedArgs;
+const backend = (supersetUrl || `http://localhost:${supersetPort}`).replace(
+  '//+$/',
+  '',
+); // strip ending backslash
+const MANIFEST_FILE = path.resolve(
+  __dirname,
+  '../superset/static/assets/manifest.json',
+);
+
+let manifestContent;
+let manifest;
+function loadManifest() {
+  try {
+    const newContent = fs.readFileSync(MANIFEST_FILE);
+    if (!newContent || newContent === manifestContent) return;
+    manifestContent = newContent;
+    manifest = JSON.parse(manifestContent);
+    console.log(`${MANIFEST_FILE} loaded.`);
+  } catch (e) {
+    if (e.code !== 'ENOENT') {
+      console.error('\n>> Error loading manifest file:');
+      console.trace(e);
+    }
+  }
+}
+// make sure the manifest file exists
+fs.mkdirSync(path.dirname(MANIFEST_FILE), { recursive: true });
+fs.closeSync(fs.openSync(MANIFEST_FILE, 'as+'));
+fs.watch(MANIFEST_FILE, loadManifest);
+
+function isHTML(res) {
+  const CONTENT_TYPE_HEADER = 'content-type';
+  const contentType = res.getHeader
+    ? res.getHeader(CONTENT_TYPE_HEADER)
+    : res.headers[CONTENT_TYPE_HEADER];
+  return contentType.includes('text/html');
+}
+
+function toDevHTML(origHtml) {
+  let html = origHtml.replace(
+    /(<head>\s*<title>)([\s\S]*)(<\/title>)/i,
+    '$1[DEV] $2 $3',
+  );
+  // load manifest file only when needed
+  if (!manifest) {
+    loadManifest();
+  }
+  if (manifest) {
+    // replace bundled asset files, HTML comment tags generated by Jinja macros
+    // in superset/templates/superset/partials/asset_bundle.html
+    html = html.replace(
+      /<!-- Bundle (css|js) (.*?) START -->[\s\S]*?<!-- Bundle \1 \2 END -->/gi,
+      (match, assetType, bundleName) => {
+        if (bundleName in manifest.entrypoints) {
+          return `<!-- DEV bundle: ${bundleName} ${assetType} -->${(
+            manifest.entrypoints[bundleName][assetType] || []
+          )
+            .map(chunkFilePath => {
+              if (assetType === 'css') {
+                return `<link rel="stylesheet" type="text/css" href="${chunkFilePath}" />`;
+              }
+              return `<script src="${chunkFilePath}"></script>`;
+            })
+            .join('\n  ')}<!-- DEV bundle: ${bundleName} ${assetType} END -->`;
+        }
+        return match;
+      },
+    );
+  }
+  return html;
+}
+
+function copyHeaders(origRes, res) {
+  res.statusCode = origRes.statusCode;
+  res.statusMessage = origRes.statusMessage;
+  if (res.setHeader) {
+    const resIsHTML = isHTML(origRes);
+    for (const [key, value_] of Object.entries(origRes.headers)) {
+      let value = value_;
+      if (resIsHTML) {
+        if (key === 'content-encoding' || key === 'content-length') {
+          continue;
+        }
```

Comment:
Keep all original response headers, but skip `content-encoding` and `content-length` for HTMLs because the modified dev HTML will not be gzipped.

Reply:
I'm not sure about this one... This is definitely more concise, but it's also three loops instead of one. Plus I liked the explicitness and directness of the naive for loops and continue/breaks.

Reply:
```js
    Object.keys(origRes.headers)
      .filter(
        key =>
          !resIsHTML || key !== 'content-encoding' || key !== 'content-length',
      )
      .forEach(key => {
        let value = origRes.headers[key];
        if (key === 'set-cookie') {
          // remove cookie domain
          value = Array.isArray(value) ? value : [value];
          value = value.map(x => x.replace(/Domain=[^;]+?/i, ''));
        } else if (key === 'location') {
          // set redirects to use local URL
          value = (value || '').replace(backend, '');
        }
        res.setHeader(key, value);
      });
```

This is probably good enough.